crates/block-producer/src/mempool/state_dag.rs

```diff
@@ -0,0 +1,442 @@
+
+impl StateDag {
+    /// Creates an empty [`StateDag`] with the given configuration.
+    pub fn new(
+        chain_tip: BlockNumber,
+        state_retention: NonZeroUsize,
+        expiration_slack: u32,
+    ) -> Self {
+        Self {
+            state_retention,
+            expiration_slack,
+            nodes: Nodes::default(),
+            state: InflightState { chain_tip, ..Default::default() },
+        }
+    }
+
+    /// Appends the transaction to the state graph if possible.
+    pub fn append_transaction(
+        &mut self,
+        tx: Arc<AuthenticatedTransaction>,
+    ) -> Result<(), AddTransactionError> {
+        self.authentication_staleness_check(tx.authentication_height())?;
+        self.expiration_check(tx.expires_at())?;
+
+        let tx_id = tx.id();
+        let node_id = NodeId::Transaction(tx.id());
+        let node = TransactionNode(tx);
+
+        // TODO: check state is available
+
+        self.nodes.txs.insert(tx_id, node);
+        self.insert_into_state_unchecked(node_id);
+
+        Ok(())
+    }
+
+    /// Selects a set of transactions for inclusion in a batch within the constraints of the budget.
+    ///
+    /// Returns [`None`] if no transactions are available for selection.
+    pub fn propose_batch(
+        &mut self,
+        mut budget: BatchBudget,
+    ) -> Option<(BatchId, Vec<Arc<AuthenticatedTransaction>>)> {
+        // The selection algorithm is fairly neanderthal in nature.
+        //
+        // We iterate over all transaction nodes, each time selecting the first transaction which
+        // has no parent nodes that are unselected transactions. This is fairly primitive, but
+        // avoids the manual bookkeeping of which transactions are selectable.
+        //
+        // This is still reasonably performant given that we only retain unselected transactions as
+        // transaction nodes i.e. selected transactions become batch nodes.
+        //
+        // The additional bookkeeping can be implemented once we have fee related strategies. KISS.
+
+        let mut selected = Vec::new();
+
+        'outer: loop {
+            'next_candidate: for candidate in self.nodes.txs.values().map(|tx| &tx.0) {
+                if selected.contains(&candidate.id()) {
+                    continue;
+                }
+
+                let parents = self.parents(NodeId::Transaction(candidate.id()));
+                for parent in parents {
+                    match parent {
+                        NodeId::Transaction(parent) if !selected.contains(&parent) => {
+                            continue 'next_candidate;
+                        },
+                        _ => {},
+                    }
+                }
+
+                if budget.check_then_subtract(candidate.as_ref()) == BudgetStatus::Exceeded {
+                    break 'outer;
+                }
+
+                selected.push(candidate.id());
+                continue 'outer;
```

Comment:
This continue outer is a bit confusing, since it only works due to the `continue block in 189, using an `let mut ier = self.nodes.txs.values().map(..);` and re-using that might be easier to understand.